cardano-testnet/src/Testnet/Components/Query.hs

```diff
@@ -364,65 +459,43 @@ getCurrentEpochNo epochStateView = withFrozenCallStack $ do
+assertNewEpochState epochStateView sbe maxWait lens expected = withFrozenCallStack $ do
+  mStateView <- watchEpochStateUpdate epochStateView maxWait (const checkEpochState)
+  when (isNothing mStateView) $ do
+    val <- getFromEpochStateForEra
+    if val == expected
+    then pure ()
+    else H.failMessage callStack $ unlines
+           [ "assertNewEpochState: expected value not reached within the time frame."
+           , "Expected value: " <> show expected
+           , "Actual value: " <> show val
+           ]
   where
-    checkEpochState :: HasCallStack
-          => ShelleyBasedEra era -> AnyNewEpochState -> m (Maybe ())
-    checkEpochState sbe newEpochState = do
-      val <- getFromEpochState sbe newEpochState
-      return $ if val == expected then Just () else Nothing
-
-    getFromEpochState :: HasCallStack
-              => ShelleyBasedEra era -> AnyNewEpochState -> m value
-    getFromEpochState sbe (AnyNewEpochState actualEra newEpochState) = do
-      Refl <- either error pure $ assertErasEqual sbe actualEra
-      return $ newEpochState ^. lens
+    checkEpochState
+      :: HasCallStack
+      => m (Maybe ())
+    checkEpochState = withFrozenCallStack $ do
+      val <- getFromEpochStateForEra
+      pure $ if val == expected then Just () else Nothing
+
+    getFromEpochStateForEra
+      :: HasCallStack
+      => m value
+    getFromEpochStateForEra = withFrozenCallStack $ getEpochStateDetails epochStateView $
+      \(AnyNewEpochState actualEra newEpochState, _, _) -> do
+        Refl <- H.leftFail $ assertErasEqual sbe actualEra
+        pure $ newEpochState ^. lens
 
--- | Watch the epoch state view until the guard function returns 'Just' or the timeout epoch is reached.
--- Wait for at most @maxWait@ epochs.
--- The function will return the result of the guard function if it is met, otherwise it will return @Nothing@.
-watchEpochStateView
-  :: forall m a. (HasCallStack, MonadIO m, MonadTest m, MonadAssertion m)
-  => EpochStateView -- ^ The info to access the epoch state
-  -> (AnyNewEpochState -> m (Maybe a)) -- ^ The guard function (@Just@ if the condition is met, @Nothing@ otherwise)
-  -> EpochInterval -- ^ The maximum number of epochs to wait
-  -> m (Maybe a)
-watchEpochStateView epochStateView f (EpochInterval maxWait) = withFrozenCallStack $ do
-  AnyNewEpochState _ newEpochState <- getEpochState epochStateView
-  let EpochNo currentEpoch = L.nesEL newEpochState
-  go (EpochNo $ currentEpoch + fromIntegral maxWait)
-    where
-      go :: EpochNo -> m (Maybe a)
-      go (EpochNo timeout) = do
-        epochState@(AnyNewEpochState _ newEpochState') <- getEpochState epochStateView
-        let EpochNo currentEpoch = L.nesEL newEpochState'
-        condition <- f epochState
-        case condition of
-          Just result -> pure (Just result)
-          Nothing -> do
-            if currentEpoch > timeout
-              then pure Nothing
-              else do
-                H.threadDelay 10_000
```

Comment:
@palas why we were using 10ms polling interval?

Reply:
Feel free to change it. My reasoning was, computers do more than 10^9 operations per second so doing 100 polls per second seems that it is not a lot of processing wasted waiting (less than 0.00001%), and at the same time the latency is almost imperceptible for humans (100th of a second)